action/protocol/poll/governance_protocol.go

```diff
@@ -303,3 +379,38 @@ func (p *governanceChainCommitteeProtocol) getGravityHeight(ctx context.Context,
 	)
 	return p.electionCommittee.HeightByTime(blkTime)
 }
+
+func (p *governanceChainCommitteeProtocol) getKickoutBlackList(epochNum uint64) (map[string]bool, error) {
+	kickoutList, err := p.kickoutListByEpoch(p.sr, epochNum)
+	if err != nil {
+		return nil, err
+	}
+	kickoutMapping := make(map[string]bool, 0)
+	for _, addr := range kickoutList {
+		kickoutMapping[addr] = true
+	}
+	return kickoutMapping, nil
+}
+
+func (p *governanceChainCommitteeProtocol) calculateKickoutBlackList(
+	ctx context.Context,
+	epochNum uint64,
+) ([]string, error) {
+	var unqualifiedDelegates []string
+	var i uint64
+	for i = 1; i <= p.kickoutEpochPeriod; i++ {
+		// check N-1, N-2, ... N-K epochs
+		numBlks, produce, err := p.productivityByEpoch(ctx, epochNum-i)
+		if err != nil {
+			return nil, err
+		}
+		expectedNumBlks := numBlks / uint64(len(produce))
+		for addr, actualNumBlks := range produce {
+			if actualNumBlks*100/expectedNumBlks < p.productivityThreshold {
+				unqualifiedDelegates = append(unqualifiedDelegates, addr)
+			}
+		}
+	}
```

Comment:
blacklist[N] = blacklists[N-1] - blacklists[N-K] + newly blocked, more efficient, and prevent two issues:
1. K is too big
2. if N = English, the blacklist is supposed to be empty, but it will recalculate the blacklist from the previous K epoch, and...

Reply:
If we change to this, the time complexity will be more efficient! From what I implemented, the calculation of blacklist has been going regardless of English height and kick-out after english height. But then the calculation of blacklist will start from EnglishHeight which is 0 at that English epoch, is that right?

Reply:
yes, does it make sense to you?

Reply:
yes, implement it on latest commit